File: 2020/app/src/main/scala/org/snowi/aoc2020/puzzle2/Puzzle2.scala
/* (C) stefan.devops@gmail.com 2021 */
package org.snowi.aoc2020.puzzle2

object Puzzle2 {

  val IDX_LOWER_TOKEN = 0
  val IDX_UPPER_TOKEN = 1
  val IDX_CHAR_TOKEN = 2
  val IDX_PW_TOKEN = 3

  def main(args: Array[String]): Unit = {
    answer()
  }

  def answer(): Unit = {
    val input = readPwFile("puzzle2/input.txt")
    println()
    println("--------------- Puzzle 02 ---------------")
    println("Result puzzle 2 part 1: " + checkPwEntries(input))
    println("Result puzzle 2 part 2: " + checkPwEntriesV2(input))
    println()
  }

  def checkPolicy1(pwEntry: String): Boolean = {
    val tokens = pwEntry.replace("-", " ").split(" ")
    val pw = tokens(IDX_PW_TOKEN)
    val lower = tokens(IDX_LOWER_TOKEN).toInt
    val upper = tokens(IDX_UPPER_TOKEN).toInt
    val c = tokens(2).replace(":", "").charAt(0)
    var i = 0
    var matches = 0

    while (i < pw.length) {
      if (pw.charAt(i) == c) {
        matches += 1
      }
      i += 1
    }
    lower <= matches && matches <= upper
  }

  def checkPolicy2(pwEntry: String): Boolean = {
    val tokens = pwEntry.replace("-", " ").split(" ")
    val pw = tokens(IDX_PW_TOKEN)
    val lower = tokens(IDX_LOWER_TOKEN).toInt - 1
    val upper = tokens(IDX_UPPER_TOKEN).toInt - 1
    val c = tokens(IDX_CHAR_TOKEN).replace(":", "").charAt(0)

    (pw.charAt(lower) == c && pw.charAt(upper) != c) || (pw.charAt(
      lower
    ) != c && pw.charAt(upper) == c)
  }

  def readPwFile(filename: String): List[String] = {
    val bufferedSource = io.Source.fromResource(filename)
    val lines = (for (line <- bufferedSource.getLines()) yield line).toList
    bufferedSource.close
    lines
  }

  def checkPwEntries(entries: List[String]): Int = {
    var correct = 0
    for (pwEntry <- entries if checkPolicy1(pwEntry)) {
      correct += 1
      // println(pwEntry)
    }

    correct
  }

  def checkPwEntriesV2(entries: List[String]): Int = {
    var correct = 0
    for (pwEntry <- entries if checkPolicy2(pwEntry)) {
      correct += 1
      // println(pwEntry)
    }

    correct
  }
}

File: 2020/app/src/main/java/org/snowi/aoc2020/puzzle2/Puzzle2.java
package org.snowi.aoc2020.puzzle2;

import org.snowi.aoc2020.Puzzle;

import java.util.List;

public class Puzzle2 extends Puzzle {
    private static final int IDX_LOWER_TOKEN = 0;
    private static final int IDX_UPPER_TOKEN = 1;
    private static final int IDX_CHAR_TOKEN = 2;
    private static final int IDX_PW_TOKEN = 3;

    public Puzzle2(String inputFile) {
        super(inputFile);
    }

    @Override
    public String answer1(List<String> input) {
        int correct = 0;
        for (String password : input){
            if(checkPolicy1(password)){
                correct++;
            }
        }
        return String.valueOf(correct);
    }

    @Override
    public String answer2(List<String> input) {
        int correct = 0;
        for (String password : input){
            if(checkPolicy2(password)){
                correct++;
            }
        }
        return String.valueOf(correct);
    }

    boolean checkPolicy1(String pwEntry) {
        String[] tokens = pwEntry.replace("-", " ").split(" ");
        String pw = tokens[IDX_PW_TOKEN];
        int lower = Integer.parseInt(tokens[IDX_LOWER_TOKEN]);
        int upper = Integer.parseInt(tokens[IDX_UPPER_TOKEN]);
        char c = tokens[IDX_CHAR_TOKEN].replace(":", "").charAt(0);
        int i = 0;
        int matches = 0;

        while (i < pw.length()) {
            if (pw.charAt(i) == c) {
                matches += 1;
            }
            i += 1;
        }
        return (lower <= matches) && (matches <= upper);
    }

    boolean checkPolicy2(String pwEntry) {
        String[] tokens = pwEntry.replace("-", " ").split(" ");
        String pw = tokens[IDX_PW_TOKEN];
        int lower = Integer.parseInt(tokens[IDX_LOWER_TOKEN]) - 1;
        int upper = Integer.parseInt(tokens[IDX_UPPER_TOKEN]) - 1;
        char c = tokens[IDX_CHAR_TOKEN].replace(":", "").charAt(0);


        return ((pw.charAt(lower) == c && pw.charAt(upper) != c) || (pw.charAt(
            lower
        ) != c && pw.charAt(upper) == c));
    }


}
